### PR TITLE
MM-52438 Clean up web app CI and only run when web app code is changed

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
This is a reworked version of https://github.com/mattermost/mattermost-server/pull/23097 after coming to the conclusion that making the pipeline more granular wasn't working out. I still took the opportunity though to clean up the pipeline by doing the following:
1. The most important one is that PRs will only run the web app CI pipeline if they actually make web app changes. I'm hoping we can do the same with the server PR soon
2. The unit tests now run in parallel with linting since linting failures don't always lead to test failures
3. Finally, I just did some general cleanup. I'm hoping that the stuff I read on "composite actions" works since I was annoyed that we kept repeating the node and NPM setups

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-52438

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: .github/actions/webapp-setup/action.yml
Comment (M-ZubairAhmed) [inline .github/actions/webapp-setup/action.yml:5]: 🎉

**Review by crspeller — APPROVED** [review]

File: .github/workflows/webapp-ci-pr.yml
Comment (agnivade) [inline .github/workflows/webapp-ci-pr.yml:5]: `**` already matches everything. Do we need to add a `/*` after that?

Comment (hmhealey) [reply .github/workflows/webapp-ci-pr.yml:5]: Good catch. I think that's left over from me trying to do something like `**/*.js`

**Comment by agnivade** [comment]
Let's merge this if everything looks good?

**Comment by hmhealey** [comment]
@agnivade Yeah, if you wouldn't mind throwing a second approval on it so that we can merge it, I'd be good with that to unblock your PR

@phoinixgrr Let us know if you have any concerns about this. I don't think this should affect your team, but I wanted to make sure someone on your team had a heads up that I was changing this

**Review by phoinixgrr — COMMENTED** [review]
Fully aligned with what proposed here. 🎖️ 
Great work! 

Especially liked the use of local composite action `.github/actions/webapp-setup/action.yml` 
 keeping the code DRY


What do you say locating the file at : `.github/actions/webapp/setup.yml` instead? 
Would it scale better in case other products, choose to DRY-up their code?  `.github/actions/{product}/{action_name}.yml`

**Review by phoinixgrr — COMMENTED** [review]
Fully aligned with what proposed here. 🎖️ 
Great work! 

Especially liked the use of local composite action `.github/actions/webapp-setup/action.yml` 
 keeping the code DRY


What do you say locating the file at : `.github/actions/webapp/setup.yml` instead? 
Would it scale better in case other products, choose to DRY-up their code?  `.github/actions/{product}/{action_name}.yml`

**Comment by hmhealey** [comment]
Sorry, I missed your comment before merging this. If I remember correctly, GitHub requires the action to specifically be named `action.yml`. We could probably change it to `.github/actions/webapp/setup/action.yml` if we want, but I think we can wait until we get more of these composite actions to start grouping them together more.

And also as a related comment, I wish I could've added the checkout step to the shared action, but since it needs to check out the repo to know what to do, I had to separate out the checkout step

**Pull request merged**
